packages/playground/src/screens/connection.tsx

```diff
@@ -44,35 +44,36 @@ function ActiveConnectionScreen({ connection }: { connection: Connection }) {
     return null
   }
 
+  useEffect(() => {
```

Comment:
> Also it seems possible that the events would be fired multiple times

Yeah, this is possible. In which case, it is theoretically possible a navigation would happen twice, if you render a Provider component more than once.

However, in practice, we would have Rise-specific (because those navigation events are exclusive to Rise) "NavigationProvider" that would be responsible to handle navigation-only events.

We would then, create providers / specialised components for each type of events (e.g. "analytics" or "logging"), making the design overall modular.

Reply:
> the effect is not as clean as the previous implementation, in my opinion

We could theoretically keep it within `<Template />` component, but I consider this an unnecessary API since `onEvent` is not tied to Template component, but to DataSource (in this particular PR).

Happy to discuss, also, maybe we can come up with a nice hook such as `useTemplateEvent(dataSource)` or something.

Reply:
As discussed, we will be returning back to previous implementation to avoid combining data source with client-side only events.